### DISCUSSION TITLE
132

### DISCUSSION THREAD

**Discussion by jojoldu** [jojoldu/blog-comments#726] [Comments]
http://jojoldu.tistory.com/132

**Comment by jojoldu** [comment]
작성자: 크고하얀곰

감사합니다. 도움이 많이 된거같아요.

**Comment by jojoldu** [comment]
작성자: 김형근

좋은 글 감사합니다. ^_^ WARN 메세지 보고 한참 찾아봤네요 ㅠㅠ

**Comment by kimsangkyun** [comment]
open() 오버라이딩 한 함수에서 doOpen()을 호출하나요??

**Comment by hongyun0** [comment]
감사합니다 ㅠ_ㅠ

**Comment by traeper** [comment]
좋은 글 감사합니다. :)

**Comment by melpis** [comment]
오타 제보합니다~!
org.springframework.batch.item.ItemReader는 인터페이스입니다. 구현 클래스는 어노테이션 기반 listner 구성에 대해 실행되지 않습니다. @Bean 메소드에서 @StepScope를 사용하는 경우 listner 어노테이션을 사용할 수 있도록 구현 클래스를 리턴해야합니다. 

listner > listener

**Comment by benelog** [comment]
저도 비슷한 내용을 작성한 후에 다시 이글을 보니까 제가 빠뜨린 부분이 보이네요.

그런데 아래 부분에 약간 의문이 듭니다..

> 즉, reader()에서 ItemReader 타입을 리턴할 경우 @StepScope의 proxyMode = ScopedProxyMode.TARGET_CLASS로 인해서 ItemReader 인터페이스의 프록시 객체를 생성하여 리턴하게 됩니다.

`ScopedProxyMode.TARGET_CLASS` 의 JAVADOC에는 아래와 같이 되어 있습니다.

```java
	/**
	 * Create a class-based proxy (uses CGLIB).
	 */
	TARGET_CLASS
```

 `@StepScope의 proxyMode = ScopedProxyMode.TARGET_CLASS` 로 '인해서' ItemReader 인터페이스의 프록시 객체를 생성하여 리턴되었다기 보다는, 그럼에도 불구하고 인터페이스의 프록시 객체가 생성되었다고 해석해야되지 않을까합니다.

`TARGET_CLASS` 로 되어 있어서 클래스타입으로 bean을 등록할때 의도하지 않게 interface 기반의 proxy가 생성되는 것을 방어하는것 같고, 이 선언이 있어도 JavaConfig에서는 `@Bean` 이 붙은 메서드의 반환형이 인터페이스라면  Proxy도 인터페이스 타입으로 생성하겠다는 의도로 파악하고, `ScopedProxyMode.TARGET_CLASS` 가 영향을 주지 않는것으로 우선 저는 추론하고 있습니다.

스프링 레퍼런스  가이드에서 관련 부분을 찾아보려는 중입니다.

**Comment by benelog** [comment]
레퍼런스 가이드에는 나와있지 않은데 [DefaultAopProxyFactory.createAopProxy()](https://github.com/spring-projects/spring-framework/blob/master/spring-aop/src/main/java/org/springframework/aop/framework/DefaultAopProxyFactory.java#L68)
에서 분기처리가 되고 있네요.

![image](https://user-images.githubusercontent.com/910151/107891661-6876e580-6f63-11eb-895c-e01e2c07c1b0.png)

**Comment by jojoldu** [comment]
@benelog 오 상세한 확인 감사합니다 상혁님 :) 
혹시 상혁님 파트에서 해당 부분 소개 해주시는걸까요? 제 파트에서 쓰려고 생각중이였는데 내용이 겹칠까봐서요!
